src/main/scala/viper/gobra/frontend/info/implementation/property/DependencyAnalysis.scala

```diff
@@ -0,0 +1,130 @@
+
+trait DependencyAnalysis extends BaseProperty { this: TypeInfoImpl =>
+
+  /**
+    * Checks that a global declaration is acyclic.
+    */
+  lazy val acyclicGlobalDeclaration: Property[PVarDecl] = createProperty[PVarDecl]{ c =>
+    val results = c.left.map{ l =>
+      entity(l) match {
+        case g: st.GlobalVariable =>
+          samePackageDependenciesGlobals(g) match {
+            case Right(gDeps) if gDeps.contains(g) => failedProp(s"The declaration of $l is cyclical")
+            case Right(_) => successProp
+            case Left(errs) => failedProp(errs)
+          }
+        case _: st.Wildcard =>
+          // Wildcard declaration can never be cyclical
+          successProp
+        case e => Violation.violation(s"Unexpected entity $e found")
+      }
+    }
+    PropertyResult.bigAnd(results)
+  }
+
+  /**
+    * Find all dependencies of a global variable, or an error message if the analysis cannot be carried out to
+    * completion
+    */
+  protected def samePackageDependenciesGlobals: st.GlobalVariable => Either[String, Vector[st.GlobalVariable]] = {
+    _.expOpt.toVector.map(globalsExprDependsOn).fold(Right(Vector())){
+      case (l, r) => for { lVars <- l; rVars <- r } yield lVars ++ rVars
+    }
+  }
+
+  /**
+    * Finds all global variables on which an expression depends. Returns a list of errors if dependency
+    * anlasys cannot be carried successfully.
+    * @param exp
+    * @return
+    */
+  def globalsExprDependsOn(exp: PExpression): Either[String, Vector[st.GlobalVariable]] = {
+    val enclosingPkg = tryEnclosingPackage(exp).get
+    val stack = mutable.Stack[PNode](exp)
+    var visitedEntities = Seq[st.Regular]()
+    // Stores all dependencies of a node (including itself).
+    var dependencies = Seq[st.GlobalVariable]()
+    var foundError: Option[String] = None
+
+    // According to the Go language spec:
+    //  Dependency analysis does not rely on the actual values of the variables, only on lexical
+    //  references to them in the source, analyzed transitively.
+    //  For instance, if a variable x's initialization expression refers to a function whose body
+    //  refers to variable y then x depends on y. Specifically:
+    //    - A reference to a variable or function is an identifier denoting that variable or function.
+    //    - A reference to a method m is a method value or method expression of the form t.m, where the (static)
+    //      type of t is not an interface type, and the method m is in the method set of t. It is immaterial
+    //      whether the resulting function value t.m is invoked.
+    //    - A variable, function, or method x depends on a variable y if x's initialization expression or
+    //      body (for functions and methods) contains a reference to y or to a function or
+    //      method that depends on y.
+    while (stack.nonEmpty && foundError.isEmpty) {
+      val elem = stack.pop()
+      reflexiveAllChildren(elem).foreach{
```

Comment:
Global variables cannot be ghost right now. I hope to lift this restriction in a future PR because this one is already too large. At any rate, the point is not that we can ignore ghost contexts, but instead that we must - this is not so much of a problem right now, but it will be when we lift the restriction that global variables must be provided in the correct order. So, my point is that if the order in which you declare/initialize the global variables is based on this ordering, then global variables cannot be considered for this dependency analysis because that may affect the order of initialization that Gobra picks - in other words, the initialization order of globals could depend on ghost state, violating the core principle that ghost code does not interfere with the "actual" program execution